completions/fish/wl-paste.fish

```diff
@@ -12,7 +12,39 @@ function __wayland_seats --description 'Print all wayland seats'
         set seat "seat0"
     end
     
-    echo "$seat"
+    for i in $seat
+        echo "$i"
+    end
+end
+
+function __wl_paste_types --description 'Print types with context'
+    set -l type (commandline -o)
+
+    for i in (seq (count $type))
+        if [ "$type[$i]" = '--primary' ]; or [ "$type[$i]" = '-p' ]
+            set clip "primary"
+        else if [ "$type[$i]" = '--seat' ]; or [ "$type[$i]" = '-s' ]
+            set seat $type[(math $i + 1)]
+        end
+    end
+
+    # Note fish does not handle passing unset variables
+    # to commands well, thus setting clip to "--primary" and passing
+    # that to wl-paste wont work, so if statements are used instead
+    set -l return
+    if test -n "$seat"; and test -n "$clip"
+        set return (wl-paste --seat "$seat" -p -l)
+    else if test -n "$seat"
+        set return (wl-paste --seat "$seat" -l)
+    else if test -n "$clip"
+        set return (wl-paste -p -l)
+    else
+        set return (wl-paste -l)
+    end
+
+    for i in $return
+        echo "$i"
+    end
```

Comment:
Is there a reason why you have to save the result to a variable and then loop over it and `echo` instead of just letting `wl-paste` invocations output the types?

Reply:
Complete does not space split the types when given a function; so if you manually pass "foo bar" it will return them as two separate options. However, when you pass a function that returns "foo bar" it will treat it as one type requiring actually manually echoing each one to give separate types.

Reply:
Sorry, but I still don't understand. This function does not really return the types, it outputs them to stdout, doesn't it? Then what does it matter how they are output? Or is `echo` somehow special in this regard?

If you mean space-separated vs newline-separated, then `wl-paste -l` should already print newline-separated types, just as a loop of `echo` invocations would.

Reply:
In fish echo doesn't necessarily write to stdout it is just how functions return values. So in this case `__wl_paste_types` is giving information to the complete function. It requires a for loop because newlines are automatically stripped so calling `echo (wl-paste -l)` would actually return something like `STRING COMPOUND_STRING` instead of 
`STRING`

`COMPOUND_STRING`

Reply:
To expand newlines aren't technically stripped; they are auto converted into a list structure, wherein we must iterate over every item and return it; otherwise if we just return the list it will output as one large string but formatted for example a list like this:
`STRING`

`COMPOUND_STRING`

would be outputted likes this `STRING\ COMPOUND_STRING\`

Reply:
> In fish echo doesn't necessarily write to stdout it is just how functions return values.

Could you please point me to some evidence / code / explanation of how that works?

As far as I know, at least in other shells I've worked with, functions don't return values (other than exit codes), and any info they wish to return they print to stdout, much like an external Unix executable would. This stdout is stdout of the function invocation, not necessarily of the shell itself — the shell could redirect function's stdout to a temporary buffer for more processing when doing completions, for example.

Reply:
I apologize, I misspoke; I was more referring to the fact that output isn't necessarily written to the screen, and instead captured and is then used.

Reply:
> so calling `echo (wl-paste -l)` would actually return something like `STRING\ COMPOUND_STRING\`

Maybe I should have been more clear. I don't mean to say you should call `echo (wl-paste -l)`, I'm saying why don't you just call `wl-paste -l` (with different additional flags in each of those `if` branches) and have it output its result directly to function's stdout? No need to capture and re-print it in any way.

Reply:
Yeah, no that makes perfect sense; sorry I had a case of severe tunnel vision